### PR TITLE
Fix doubly linked list

### PR DESCRIPTION
**Pull request by AugustinasJarockis**
Fixed the doubly linked list

### CODE REVIEW

File: lib/main.c [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#include "HTML_code_doubly_linked_list.h"
+
+int main()
+{
+    char *name = "html";
+    insert_tag(name, NULL);
+    char *name1 = "head";
+    insert_tag(name1, "Harambe is good");
```
Comment (arminasbrazenas) [inline lib/main.c:8]: k

File: lib/main.c [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    char *value = "Reginos receptai";
+    insert_text(value);
```
Comment (arminasbrazenas) [inline lib/main.c:10]: Maybe insert_tag and insert_text should be one function? Something like `void insert_tag (char *tag, char *params, char *text);`? Then check if text is null. Calling two functions is kind of inconvenient.

File: lib/HTML_code_doubly_linked_list.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        printf("Failed to allocate memory for end_text\n");
+        return;
+    }
+    strcat(end_text, "</");
+    strcat(end_text, name);
+    strcat(end_text, ">");
+
+    if(_HTML_START_ != NULL)
+    {
+        HTML_CODE *field_start = create_HTML_CODE(start_text, _HTML_CURRENT_, (*_HTML_CURRENT_).next);
```
Comment (arminasbrazenas) [inline lib/HTML_code_doubly_linked_list.c:84]: Unused variables

File: lib/HTML_code_doubly_linked_list.c [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        HTML_CODE *field_end = create_HTML_CODE(end_text, _HTML_CURRENT_, (*_HTML_CURRENT_).next);
+        _HTML_CURRENT_ = (*_HTML_CURRENT_).previous;
+    }
+    else
+    {
+        HTML_CODE *field_start = create_HTML_CODE(start_text, NULL, NULL);
```
Comment (arminasbrazenas) [inline lib/HTML_code_doubly_linked_list.c:90]: These three lines in if / else are really similar, could just extract it.

File: lib/HTML_code_doubly_linked_list.c [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    }
+    else
+    {
+        printf("ERROR: no code has been generated\n");
+    }
+}
+
+HTML_CODE *create_HTML_CODE(char *text, HTML_CODE *previous, HTML_CODE *next)
+{
+    HTML_CODE *new_elem = malloc(sizeof(HTML_CODE));
```
Comment (arminasbrazenas) [inline lib/HTML_code_doubly_linked_list.c:31]: There is no point of malloc?

File: lib/HTML_code_doubly_linked_list.c [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        printf("Failed to allocate memory for new element\n");
+        return NULL;
+    }
+    (*new_elem).next = next;
+    (*new_elem).previous = previous;
+    (*new_elem).text = text;
+    if(previous != NULL)
+        (*previous).next = new_elem;
+    if(next != NULL)
```
Comment (arminasbrazenas) [inline lib/HTML_code_doubly_linked_list.c:42]: Format consistently. If you are putting curly brackets on all sentences, put them on all sentences :D Also, after if's and loops there should be a space and before function calls there shouldn't be spaces. It helps to differentiate.

File: lib/HTML_code_doubly_linked_list.c [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+HTML_CODE *_HTML_START_ = NULL;
+HTML_CODE *_HTML_CURRENT_ = NULL;
+
+void generate_html_file(char *file_name)
+{
+    if(_HTML_START_ != NULL)
+    {
+        FILE *file = fopen(file_name, "w");
+        if(file == NULL) {
+            printf("Failed to open the output file\n");
```
Comment (arminasbrazenas) [inline lib/HTML_code_doubly_linked_list.c:12]: Could add [strerror ](https://www.tutorialspoint.com/c_standard_library/c_function_strerror.htm)for better experience

**Review by arminasbrazenas — CHANGES_REQUESTED** [review]

File: lib/HTML_code_doubly_linked_list.c (hunk 5 above)
Comment (AugustinasJarockis) [reply lib/HTML_code_doubly_linked_list.c:31]: There is. Then you leave the function statically allocated variables get destroyed.

File: lib/main.c (hunk 2 above)
Comment (AugustinasJarockis) [reply lib/main.c:10]: Maybe. Depends how text works and if new line matters. But you use it very rarely. I don't know why you believe merging it into one function is more convenient when you need text like 5% of time

File: lib/HTML_code_doubly_linked_list.c (hunk 3 above)
Comment (AugustinasJarockis) [reply lib/HTML_code_doubly_linked_list.c:84]: Do not really matter. But it is unnecessary.

File: lib/HTML_code_doubly_linked_list.c (hunk 7 above)
Comment (AugustinasJarockis) [reply lib/HTML_code_doubly_linked_list.c:12]: How often do you fail to open a file for writting?

File: lib/HTML_code_doubly_linked_list.c (hunk 6 above)
Comment (AugustinasJarockis) [reply lib/HTML_code_doubly_linked_list.c:42]: Seriously. Tau tarpas nepatinka. Tu checkini ar tai funkcija, ar ifas pagal ifa arba kabliataski gale

File: lib/HTML_code_doubly_linked_list.c (hunk 4 above)
Comment (AugustinasJarockis) [reply lib/HTML_code_doubly_linked_list.c:90]: Would make the functions less intuitive to use. Less code is not the only thing you should optimize for.

File: lib/main.c (hunk 2 above)
Comment (arminasbrazenas) [reply lib/main.c:10]: wdym you use it rarely? most of html elements contain text :D

File: lib/HTML_code_doubly_linked_list.c (hunk 7 above)
Comment (arminasbrazenas) [reply lib/HTML_code_doubly_linked_list.c:12]: how often do you fail to allocate memory huh?

File: lib/HTML_code_doubly_linked_list.c (hunk 3 above)
Comment (arminasbrazenas) [reply lib/HTML_code_doubly_linked_list.c:84]: cleaner code

**Review by arminasbrazenas — APPROVED** [review]

**Pull request merged**